cpp/include/Ice/Value.h

```diff
@@ -80,6 +73,15 @@ namespace Ice
 
     protected:
         /// \cond INTERNAL
+        Value(const Value&) = default; // for clone
+
+        // Helper class that allows derived classes to clone "this" even though the copy constructor is protected.
```

Comment:
Why do we need this? can't we use the protected copy constructor in clone implementation?

```
diff --git a/cpp/src/slice2cpp/Gen.cpp b/cpp/src/slice2cpp/Gen.cpp
index 318a21cdb2..0d26fb6c40 100644
--- a/cpp/src/slice2cpp/Gen.cpp
+++ b/cpp/src/slice2cpp/Gen.cpp
@@ -2473,7 +2473,7 @@ Slice::Gen::DataDefVisitor::visitClassDefEnd(const ClassDefPtr& p)
     C << sp;
     C << nl << "::std::shared_ptr<::Ice::Value>" << nl << scoped.substr(2) << "::_iceCloneImpl() const";
     C << sb;
-    C << nl << "return CloneEnabler<" << name << ">::clone(*this);";
+    C << nl << "return std::shared_ptr<" << name << ">(new " << name << "(*this));";
     C << eb;
```

Reply:
It's because we return a shared_ptr, not a raw pointer.

CloneEnabler is all about allowing make_shared_ptr to "access" the copy constructor.

Keep in mind that mapped classes should be used / created only wrapped in shared_ptr.

Reply:
Naturally, make_shared is preferred over shared_ptr(new ...) since there is a single dynamic memory allocation.